### PR TITLE
Replace instances of NULL with nullptr

### LINKED ISSUE #2345: [Feature Request] Replace instances of NULL with nullptr
**Summary**

Since we now require a C++11 compatible compiler to compile LAMMPS, we should also replace uses of `NULL` with `nullptr`.

This has two advantages
1. `nullptr` is an intrinsic element of C++11, no header needs to be included to access it.
2. Contrary to C compilers, C++ compilers define `NULL` as 0 (and not as `(void *)0`), while `nullptr` is a real pointer and thus the compiler can provide warnings if by accident there is a mismatch and one is mixing up a scalar and a pointer variable, it avoids ambiguities or unexpected choices of overloaded alternative functions and more.

**Further Information, Files, and Links**

https://embeddedartistry.com/blog/2017/03/08/migrating-from-c-to-c-null-vs-nullptr/

### PR DESCRIPTION
**Pull request by eagunn**
**Summary**

Makes the changes recommended by @akohlmey in #2345.

Changes so far have been applied to src and lib folders only.

**Related Issues**

Closes #2345 

**Author(s)**

E. Anne Gunn, Sheridan College (NWCCD)
agunn@sherprog.com

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

These changes are believed to be completely backward compatible. If they build, they should run.

**Implementation Notes**

The following changes have been applied to src and lib folders:
- regex replace: ([^"_])NULL ⇒ \1nullptr (8968 chgs in src, 1153 in lib)
- Manually find/change: (void \*) nullptr ⇒ nullptr (1 case)

The above bulk edits change the word NULL in a small number of user-visible error messages. While some of the messages might be more technically accurate using 'nullptr' rather than 'NULL', in many messages NULL is being used as a more general term for nothing or zero. I decided to revert these message changes and not alter any user-visible output. 
- regex find: ".*?nullptr.*?"
  Manually ~14 cases back to "NULL" in src, ~2 in lib
  (the regex finds a few false positive where nullptr appears between two strings in a function call)

After a successful build and regression tests, I went back and examined comments changed by batch conversion of NULL
to nullptr, I've corrected a small number where it seemed clear that the previous NULL should have been either "NULL" (where the code is looking for a string with that value),  or null char / null byte (where the comment is talking about the char/byte that terminates a c-style string). Also changed 6 places where the batch change had changed 'NULL ptr' to 'nullptr ptr'. Now they simply say 'nullptr'.

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

The original issue referenced this
[Embedded Artistry blog post](https://embeddedartistry.com/blog/2017/03/08/migrating-from-c-to-c-null-vs-nullptr/)
as a source of wisdom and background info for this change.

### CODE REVIEW

**Comment by eagunn** [comment]
@akohlmey I see the compilation_tests failure. And Jenkins shows me which build fails.
But I do not know how to get the build output to see what file/line is the problem.
Please advise.

**Comment by eagunn** [comment]
@akohlmey I have updated the comments to remove my WIP reminder and complete the implementation notes.

I am happy to do more work on this PR if you identify tasks that need to be done. But, as far as I know, this work is complete and ready for review/merge.

**Comment by akohlmey** [comment]
@eagunn thanks for your efforts. I've made some cosmetic updates in the comments, where "null pointer" would be better than "nullptr". I've also updated the legacy compilation scripts to include -std=c++11, so the changes to lib/poems and lib/awpmd could be restored. I noticed that lib/atc and lib/gpu would otherwise fail that test, too, only the legacy builds would not include those. I've changed that to include the USER-ATC package for the one compatible build.

Whether this goes in now is down to the other LAMMPS developers (it needs at least one that is not me to approve for merging).

**Comment by eagunn** [comment]
@akohlmey + Other reviewers, since this PR has not yet been ok'd, I'm going to re-apply the changes to lib/poems and lib/awpmd this morning and also attempt to match @akohlmey's clean-up of comments. Seems as if this amount of change is most useful if the result is a code base as consistent as we can make it.

**Comment by eagunn** [comment]
Looking at merge conflicts now.

**Comment by eagunn** [comment]
@akohlmey I believe I've resolved the merge conflicts. The actual fixes, one small set of lines in each of two files, were easy enough to figure out. But I'm working at the edge of my git merge/rebase skills. So I'm hoping I have not created commit/log chaos. If last change was a step backwards, I may have to ask you to correct my work.

**Comment by akohlmey** [comment]
resolved merge conflicts resulting from recent changes to upstream

**Comment by akohlmey** [comment]
@sjplimp @athomps @stanmoore1 @rbberger any comments about merging this?

**Review by athomps — APPROVED** [review]
I will use nullptr from now on.

**Comment by akohlmey** [comment]
@eagunn you rarely want to do a rebase once you have pushed a feature branch to the public. That causes all kinds of complications because it requires you to do a force push and if something fails early in the redoing the commits the effort to re-integrate changes will grow exponentially.

Better to checkout the latest master from upstream and then merge this into your feature branch. For each merge conflict, you can then do a 3-way compare and for changes like in this branch it is often the simplest to just go to the conflicted regions and take the upstream version and then manually reapply the changes from the feature branch. in this case it was a couple of clicks with a GUI merge tool and doing a query-replace of NULL to nullptr. Took only a few minutes.

File: src/write_restart.cpp [hunk 1]
```diff
@@ -661,7 +661,7 @@ void WriteRestart::write_double(int flag, double value)
 }
 
 /* ----------------------------------------------------------------------
-   write a flag and a char string (including NULL) into restart file
```
Comment (rbberger) [inline src/write_restart.cpp:664]: This is incorrect. It writes the 0 byte at the end of the string. This has nothing to do with nullptr

Comment (akohlmey) [reply src/write_restart.cpp:664]: how about this?

**Review by rbberger — APPROVED** [review]

**Comment by athomps** [comment]
@akohlmey This is a great summary of what I should always do:
> Better to checkout the latest master from upstream and then merge this into your feature branch. For each merge conflict, you can then do a 3-way compare and for changes like in this branch it is often the simplest to just go to the conflicted regions and take the upstream version and then manually reapply the changes from the feature branch.

**Comment by akohlmey** [comment]
> @akohlmey This is a great summary of what I should always do:
> 
> > Better to checkout the latest master from upstream and then merge this into your feature branch. For each merge conflict, you can then do a 3-way compare and for changes like in this branch it is often the simplest to just go to the conflicted regions and take the upstream version and then manually reapply the changes from the feature branch.

It would be great if we could have some small document collecting some of the git best practices and conventions for LAMMPS written by people that are not using git as much as I do. For me it is difficult to determine what kind of information people need and what is distracting. It is easy to make suggestions after the fact, but up-front explanations often are missing the target since my perspective on the process is different.

**Review by akohlmey — APPROVED** [review]

File: src/write_restart.cpp (hunk 1 above)
Comment (eagunn) [reply src/write_restart.cpp:664]: @rbberger This comment is an example of several I saw where the original comment used NULL rather loosely. I believe 
`write a flag and a char string (including NULL) into restart file`
should have said something more like :
`write a flag and a char string (including the terminating null char) into restart file`

My bulk change converted the NULL to nullptr in comments as well as code. Which in this case was certainly wrong-er than the original. I believe that, in the large majority of comments, the conversion to nullptr was the right choice.

When I hand-applied my cleanup to changed comments, I tried to notice the places where the original comment had used NULL in this way and replace with a comment more like the second above. I assume @akohlmey may have done the same. But, as this comment shows, our eyeballing and hand-fixing did not necessarily catch and improve every comment that used NULL in this way.

I'm happy to make another pass over comments looking for this specific problem.

Comment (rbberger) [reply src/write_restart.cpp:664]: Axel already fixed it

**Pull request merged**